src/Protocols/DfLikeAlternative.hs

```diff
@@ -0,0 +1,416 @@
+{-|
+Defines a class for protocols which are "similar" to 'Df'.
+-}
+
+{-# LANGUAGE FlexibleContexts, FlexibleInstances, MultiParamTypeClasses, NamedFieldPuns, UndecidableInstances #-}
+
+module Protocols.DfLikeAlternative where
+
+import           Prelude hiding (replicate, last, repeat, unzip)
+import           Control.Arrow ((***))
+import           Control.Monad (when)
+import           Control.Monad.State (State, runState, get, put, gets, modify)
+import           Clash.Prelude hiding ((&&), (||), not)
+import           Data.Maybe (isJust, fromJust)
+import           Data.Proxy (Proxy(..))
+
+-- me
+import           Protocols.Axi4.Common (KeepBurst(..), KeepSize(..), KeepBurstLength(..), KeepResponse(..), KeepStrobe(..), Width, BurstMode(..), Resp(..))
+import           Protocols.Axi4.ReadAddress (Axi4ReadAddress, M2S_ReadAddress(..), S2M_ReadAddress(..))
+import           Protocols.Axi4.ReadData (Axi4ReadData, M2S_ReadData(..), S2M_ReadData(..))
+import           Protocols.Axi4.WriteAddress (Axi4WriteAddress, M2S_WriteAddress(..), S2M_WriteAddress(..))
+import           Protocols.Axi4.WriteData (Axi4WriteData, M2S_WriteData(..), S2M_WriteData(..))
+import           Protocols.Axi4.WriteResponse (Axi4WriteResponse, M2S_WriteResponse(..), S2M_WriteResponse(..))
+import           Protocols.Df (Data(..), Df)
+import qualified Protocols.Df as Df
+import           Protocols.Internal
+
+
+-- | Class for protocols that are "similar" to 'Df',
+-- i.e. they can be converted into a 'Df' port using a 'Circuit'.
+-- Can take parameters, e.g. for addresses.
+-- Defaults to being the right side of the circuit (Fwd ~ Otp, Bwd ~ Inp),
+-- but this can be switched using 'Reverse' from 'Protocols.Internal'.
+-- Supports both input data and output data
+class (Protocol df) => DfLikeAlternative df where
+  -- | Domain that messages are being sent over
+  type Dom df :: Domain
+  -- | Information being sent in
+  type InpPayload df
+  -- | Information being sent out
+  type OtpPayload df
+  -- | User-provided parameters (e.g. address to respond to)
+  type DfLikeParam df
```

Comment:
I updated the comment; better now? The gist is that we want to have addresses be parameterized so that circuits can respond to different addresses. For protocols without addresses (or anything else that needs to be parameterized), DfLikeParam can be set to ()